### PR TITLE
feat: support execution of delegations

### LINKED ISSUE #195: Connection should be able to take materialized invocations
I was telling @gobengo that we could route request from one server to the other simply by passing them to `connection.execute`, but turns out this was incorrect because:

1. Connection's execute takes `IssuedInvocation` (Well tuple of `ServiceInvocation`s but those are basically `IssuedInvocation`s) which is not an `Delegation` it's like a lazy version that can be materialized as needed

   https://github.com/web3-storage/ucanto/blob/582c4c504d2e75feee2c5d298ea08b2f01ff1c5e/packages/interface/src/lib.ts#L232-L239

2. Connection passes these lazy invocations into encoder to encode

   https://github.com/web3-storage/ucanto/blob/582c4c504d2e75feee2c5d298ea08b2f01ff1c5e/packages/client/src/connection.js#L47-L49

3. Which in turn create actual delegations by materializing them

   https://github.com/web3-storage/ucanto/blob/582c4c504d2e75feee2c5d298ea08b2f01ff1c5e/packages/transport/src/car.js#L19-L23 

Few things to note here:

1. For a while I have been considering to get rid off these lazy invocation things which would make .invoke and .delegate functions more similar. I'm still hesitant however as they introduce additional awaits, first you await to create invocation then you await to execute the invocation.
2. Even with lazy invocations I think it is a mistake to make materializing them transport's problem. It would make a lot more sense to make `execute` materialize all of the invocations first and then pass those onto transport to just do encoding.
   - It probably ended up in transport because there was a loop iterating invocations and I just avoided second one

### Option 1

- I think we should change transport API so it expects materialized invocations and make materializing them connection's problem
- We could also loosen up execute function to allow passing both materalized and non-materialized invocations so it could materialize ones that aren't yet and forward them onto transport

### Option 2

Perhaps whole materialized / not-yet-materalized invocations could be abstracted away with some `EncodableInvocation` interface of sorts that provides some method e.g. `.delegate()` which returns a materialized invocation. That way connection, transport and all those layers could just expect `EncodableInvocation` and not care which one they're actually getting.

Other than type changes it would basically mean adding `delegate() { return this }` to the `Delegation` class because `Invocation` already has `.delegate()`

### PR DESCRIPTION
**Pull request by Gozala**
Fixes #195

### CODE REVIEW

**Review by gobengo — APPROVED** [review]

**Review by gobengo — APPROVED** [review]

File: packages/transport/test/car.spec.js [hunk 1]
```diff
@@ -121,6 +122,8 @@ test('accepts Content-Type as well', async () => {
     expiration,
   })
 
+  CAR.encode([delegation])
```
Comment (gobengo) [inline packages/transport/test/car.spec.js:125]: is this just asserting that `encode` doesn't throw? or does the call to `encode` have an important side effect?

Comment (Gozala) [reply packages/transport/test/car.spec.js:125]: No just a mistake, I'll remove it

**Comment by hugomrdias** [comment]
@Gozala where is this https://github.com/web3-storage/ucanto/blob/e84c659268d104e24bfc0d0354ef436b865b12ca/packages/core/src/delegation.js#L192 function called now ? This is where the ucan is signed and it does not seems to be called.

**Comment by gobengo** [comment]
@hugomrdias that `delegate` fn is called by `IssuedInvocation#delegate` https://github.com/web3-storage/ucanto/blob/e84c659268d104e24bfc0d0354ef436b865b12ca/packages/core/src/invocation.js#L52 which, after this PR, is called by the transport encoders [eg here](https://github.com/web3-storage/ucanto/pull/199/files#diff-520f4a0baea89881cb7747d65a4ad3af8144dab526c7edd4d9597995c73eb5e2R23)

**Pull request merged**